distribution/lib/Standard/Base/0.0.0-dev/src/Data/Any.enso

```diff
@@ -94,6 +94,9 @@ type Any
                Therefore, there are no more cases to handle in self method.
             _ -> False
 
+    hash_code : Integer
+    hash_code self = Meta.hash_code self
```

Comment:
This turns Enso into [hash code oriented language](http://wiki.apidesign.org/wiki/Smalltalk). That may not be a bad thing - at least @kustosz likes the way `hashCode` is handled in Java. However for API perspective:

I don't like that any type can override this method `hash_code`! That's not what we want - we'd rather guarantee immutable `hash_code`. As far as I can say, that would better be done by having the method only in `Meta.hash_code`, not here.

Reply:
I agree. What was the original reason for allowing users to override `hash_code` on their custom types? The reasons that I can think about are:

- Ensure that the type can be used as a key into a hash map. Which is already ensured by the fact that Atoms in Enso are immutable and that, by default, a hash code is computed for every Atom. So users should not have to implement their own hash code computation for this purpose.
- Performance. This should hardly be a reason for our users, given that we cache the hash code computation. And I doubt there will be any users caring about the warm-up performance so much that they would be willing to tune their own `hash_code` method.

Is there anything else that I am missing @kustosz ?

Reply:
We allow users to override `==` on custom types. If they do so, they need to also override `hash_code` to be consistent with the changed `==` semantics if the user wants to use the type in any HashMaps (even implicitly, e.g. `Vector.distinct` in the future).

Reply:
It is perfectly ok to have
```
type My_Type
    Value x y
   
    == self other =
        self.x+self.y == other.x+other.y
```

And then have `My_Type.Value 1 2 == My_Type.Value 2 1` - in fact we have some tests like that somewhere with custom comparators.

Now, with that changed `==`, I will need to adapt `hash_code` too in order to be consistent.

Reply:
This poses another question - possibly for later - but maybe we should consider to somehow warn the user if they override `==` but do not override `hash_code` as they can get weird results otherwise.

Reply:
Oh, right. I completely forgot about `==` operator :-). Then I guess we should let users override `hash_code` indeed.

Reply:
> warn the user if they override `==` but do not override `hash_code` as they can get weird results otherwise.

Enforcing consistency between `hash_code` and `==` is important. Java completely fails in that (because it doesn't separate [API from SPI](http://wiki.apidesign.org/wiki/APIvsSPI) of course ;-). Scala is probably better - I've heard it enforces symmetry of `==`.

If we make `==` a special operator that cannot be overriden (calling `Meta.is_equal_object`) we could  enforce not only symmetry, but also check consistency of `hash_code`.

Reply:
I think we could just disallow overriding a particular method, like `==` in a compiler pass.

The question (if I understood you correctly) is different: do you think we really want to completely disallow the user from overriding `==`?

We definitely want in some cases to be able to introduce custom notions of equality. For example I may want to count how many distinct elements up-to some isomorphism a collection has.

But maybe indeed it can be good-enough (or even: better!) to achieve this kind of functionality not through allowing to override `==` but through some other interface for defining custom equality constraints? For example such interface could require a custom hashing predicate too, making it much more clear that hashcode is required for a given operation and making it more likely the user will ensure the consistency between the two.

I guess we already have _some_ notion similar to this - in `Vector.sort` we have the `by` parameter which allows specifying a custom comparator - thus the user does not need to wrap values into a custom type overriding `.compare_to` but can simply provide a custom comparator. Maybe a similar solution could be used for example for our Hash_Map or for `Vector.distinct` (where we would have some `Equality` type, defaulting to a default non-overridable `==`+`hash_code` pair, but possibly being able to specify another kind of a predicate there).

The more I write about this idea the more I like it - it makes controlling the custom notions of equality much more explicit and seems to make life much safer. Then the default equality would be not overridable and thus we'd enforce its correctness.

I think I see 2 potential disadvantages:
1. Algorithms need to be explicitly parameterized with custom equality. If a user creates an algorithm relying on `==`, there's no need to override the notion of equality for it. That may not necessarily be a bad thing though, IMO this kind of explicitness is actually better. But users may not think about it and make algorithms that are less customizable (OTOH I prefer that than having algorithms that will become incorrect because someone thrashed how equality works).
  i. The above may be especially problematic for polyglot - if I want to pass my Enso objects to some Java HashMap, there I won't have my customizable notion of equality so I may be stuck with the default always. This may not be the end of the world, because we could provide some 'magic' polyglot wrapper which wraps the Enso object and adds the custom notion of equality. But that then would essentially allow the override just in a weird advanced way so I'm not sure if that would be the way to go.
2. _I forgot what the other one was :grin: if it comes back I will add it_

Reply:
> The more I write about this idea the more I like it - it makes controlling the custom notions of equality much more explicit and seems to make life much safer. Then the default equality would be not overridable and thus we'd enforce its correctness.

If we want `hash_code`/`==` for all Enso objects, then _default equality_ is easiest way to get there. Giving advanced users access to their own `Map` with _custom equality_ is probably the best way to go for cases where the default isn't enough.

> I want to pass my Enso objects to some Java HashMap

I guess you should create the Enso `Map` (with default or custom equality) and pass it to Java.

>  forgot what the other one was

Here is one. How shall `==` behave for `Map` (with/without custom equality)?